[1/3] SourceLocation related refactoring.

christianparpart opened the pull request:
This pr is the earliest in the series of PRs that needs to be merged "in order" :-).

**Motivation:** I noticed, that we pass (CharStream, filename) almost always as pair to Scanner, and then use complex ways (via callback functions) to get back to some functions that can extract the source code fragments (duplicated API from CharStream into Scanner) inside the SourceReferenceFormatter.

**Gains**:
* So the idea now is, to not pass CharStream **and** a source name to the scanner, but to simply pass a CharStream, that already knows the source name, to the Scanner.
* This should eliminate the callback workaround as SourceLocation will have access to the CharStream already.

Prerequisite of: #5535 (is based on top of this PR)

Commit:
liblangutil: extends CharStream to know about the respective (file-)name (and adapt codebase to it)

Commit:
liblangutil: adds ParserBase::charStream() accessor

Commit:
liblangutil: SourceLocation: adds (shared) pointer to underlying CharStream source, eliminating sourceName

Also, adapted affecting code to those changes.

Commit:
liblangutil: adds Scanner.charStream() accessor

Commit:
liblangutil: Scanner: remove superfluous sourceName field (it's in CharStream already)

Also, ParserBase::sourceName() was dead code. Eliminating it should
increase test coverage (how sneaky) :-)

chriseth commented on liblangutil/ParserBase.h:
```diff
@@ -38,6 +39,7 @@ class ParserBase
 	explicit ParserBase(ErrorReporter& errorReporter): m_errorReporter(errorReporter) {}
 
 	std::shared_ptr<std::string const> const& sourceName() const;
+	CharStream const* charStream() const { return &m_scanner->charStream(); }
```
Why not return by reference?

chriseth replied:
Sorry, I don't remember the conclusion of our discussion here. Why is it fine to return by pointer here but return by reference in the scanner?

christianparpart replied:
It's returning a `shared_ptr<CharStream>` now. I wonder why Github isn't showing it here.

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -67,37 +68,46 @@ struct SourceLocation
 	int start;
 	int end;
-	std::shared_ptr<std::string const> sourceName;
+	CharStream const* charStream; // non-owning, readonly
```
I'm not sure if this is a good idea. With this change, all instances of SourceReference rely on the underlying CharStream's lifetime.

christianparpart replied:
Discussion resolution: use `shared_ptr<CharStream>`

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -67,37 +68,46 @@ struct SourceLocation
+
+	std::string const& sourceName() const
+	{
+		static std::string const empty;
```
This feels inconsistent with regards to `hasSourceName`.

christianparpart replied:
I'm open for suggestions. I did not want to return a `shared_ptr` or const pointer. that's a common trick I've been used to (although, not fixated to) -> boost::optional<std::string> would be the better fit then, IMHO.

chriseth commented on liblangutil/Scanner.h:
```diff
@@ -159,8 +157,8 @@ class Scanner
-		solAssert(m_sourceName && _location.charStream, "");
-		solAssert(*m_sourceName == _location.charStream->filename(), "");
+		solAssert(!m_source.filename().empty() && _location.hasSourceName(), "");
+		solAssert(m_source.filename() == _location.sourceName(), "");
```
You could even assert that the memory location of the char stream is the same.

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -37,28 +38,28 @@ namespace langutil
 	{
 		return start == _other.start && end == _other.end &&
-			((!sourceName && !_other.sourceName) || (sourceName && _other.sourceName && *sourceName == *_other.sourceName));
+			((!source && !_other.source) || (source && _other.source && source->name() == _other.source->name()));
```
I would prefer `source.get() == _other.source.get()` - it would be more accurate with regards to the `--CODEGEN--`-sources. Then of course this whole expression can be simplified.

christianparpart replied:
okay, thx.

chriseth commented on libevmasm/Assembly.cpp:
```diff
@@ -186,11 +186,11 @@ class Functionalizer
 		m_out << m_prefix << "    /*";
-		if (m_location.sourceName)
-			m_out << " \"" + *m_location.sourceName + "\"";
+		if (m_location.source->name() != "")
```
I think the empty source name should not receive any special treatment. That is what we have the empty source pointer for.

chriseth commented on liblangutil/Scanner.cpp:
```diff
@@ -169,15 +169,21 @@ class LiteralScope
 
+void Scanner::reset(std::shared_ptr<CharStream> _source)
+{
+	m_source = _source;
```
Does it make sense to move here?

chriseth commented on liblangutil/Scanner.cpp:
```diff
@@ -169,15 +169,21 @@ class LiteralScope
 {
-	m_source.reset();
-	m_char = m_source.get();
+	m_source->reset();
```
Is it guaranteed to be non-null?

chriseth replied:
Ah, it more or less is.

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -69,35 +70,35 @@ struct SourceLocation
 /// Stream output for Location (used e.g. in boost exceptions).
 inline std::ostream& operator<<(std::ostream& _out, SourceLocation const& _location)
 {
 	if (_location.isEmpty())
```
Or source is null.

christianparpart replied:
handled a line below (now).

chriseth commented on libsolidity/ast/ASTJsonConverter.cpp:
```diff
@@ -122,8 +122,8 @@ void ASTJsonConverter::setJsonNode(
 	int sourceIndex{-1};
-	if (_location.sourceName && m_sourceIndices.count(*_location.sourceName))
-		sourceIndex = m_sourceIndices.at(*_location.sourceName);
+	if (_location.source && _location.source->name() != "" && m_sourceIndices.count(_location.source->name()))
```
The non-empty check was not part of the original version.

chriseth commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -922,8 +922,8 @@ string CompilerStack::computeSourceMapping(eth::AssemblyItems const& _items) con
 		int sourceIndex =
-			location.sourceName && sourceIndicesMap.count(*location.sourceName) ?
-			sourceIndicesMap.at(*location.sourceName) :
+			location.source && location.source->name() != "" && sourceIndicesMap.count(location.source->name()) ?
```
Also here, please remove the non-empty check.

christianparpart replied:
Thanks for finding these.

chriseth commented on test/libsolidity/Assembly.cpp:
```diff
@@ -132,12 +132,13 @@ void printAssemblyLocations(AssemblyItems const& _items)
 		if (_items[i].location() != _locations[i])
 		{
-			BOOST_CHECK_MESSAGE(false, "Location mismatch for item " + to_string(i) + ". Found the following locations:");
+			// TODO this test is TEMPORARILY disabled until I've found a proper way to fix the part below
```
Needs to be re-activated before merge.

christianparpart replied:
done.

chriseth commented on test/libsolidity/Assembly.cpp:
```diff
@@ -151,29 +152,39 @@ BOOST_AUTO_TEST_SUITE(Assembly)
+
+	// CharStream const codegenCharStream { "", "--CODEGEN--" };
+	//
+	// ATTENTION:
```
Probably best to relax the test to only check for start/end.

christianparpart replied:
I've found a way to do that. thanks :-)

chriseth commented on libevmasm/Assembly.cpp:
```diff
@@ -117,7 +117,7 @@ string locationFromSources(StringMap const& _sourceCodes, SourceLocation const&
 		return "";
 
-	auto it = _sourceCodes.find(*_location.sourceName);
+	auto it = _sourceCodes.find(_location.source->name());
```
Please protect against `_location.source` being null - I know it was not there before, but don't let bad legacy code stand in our way :)

christianparpart replied:
done 2 lines above (now). thx.

chriseth commented on libevmasm/Assembly.cpp:
```diff
@@ -186,11 +186,10 @@ class Functionalizer
 	void printLocation()
 	{
-		if (!m_location.sourceName && m_location.isEmpty())
+		if (!m_location.source && m_location.isEmpty())
```
Shouldn't this be `||`?

christianparpart replied:
actually yes. I feel like i'm not concentrated, sorry.

chriseth commented on libevmasm/Assembly.cpp:
```diff
@@ -186,11 +186,10 @@ class Functionalizer
+		if (!m_location.source && m_location.isEmpty())
 			return;
 		m_out << m_prefix << "    /*";
-		if (m_location.sourceName)
```
Please re-add the check: `m_location.source`

chriseth commented on liblangutil/Scanner.h:
```diff
@@ -90,14 +90,16 @@ class Scanner
 {
 	friend class LiteralScope;
 public:
+	explicit Scanner(std::shared_ptr<CharStream> _source) { reset(std::move(_source)); }
```
Please assert that `_source` is not null. We dereference it all the time in the code without checks.

christianparpart replied:
the Ctor is calling `reset(sourcePtr)`; I've added the assert to the implementation of `reset(sourcePtr)` to make sure we catch it more generically. (If one wishes to fully reset (instead of re-initialize), they can use `reset()`).

chriseth commented on liblangutil/Scanner.h:
```diff
@@ -90,12 +90,16 @@ class Scanner
+	explicit Scanner(CharStream _source = CharStream()) { reset(std::move(_source)); }
 
-	std::string source() const { return m_source.source(); }
+	std::string source() const { return m_source->source(); }
```
Should perhaps return by `const&`?

christianparpart replied:
I noticed that too. But I didn't want to change too much that's not explicitly related to SourceLocation->CharStream refactor. Since seperate PR for smallies like that don't make much sense, I'll do that.

Commit:
liblangutil: small refactors wrt. API cleanups

Especially also remove SourceLocation ctor's that the compiler can default-implement.

chriseth commented on libevmasm/Assembly.cpp:
```diff
@@ -186,11 +186,11 @@ class Functionalizer
 	void printLocation()
 	{
-		if (!m_location.sourceName && m_location.isEmpty())
+		if (!m_location.source || m_location.isEmpty())
```
Actually the `&&` was correct - there are individual checks below...

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -37,28 +38,29 @@ namespace langutil
 		return start == _other.start && end == _other.end &&
-			((!sourceName && !_other.sourceName) || (sourceName && _other.sourceName && *sourceName == *_other.sourceName));
+			((!source.get() && !_other.source.get()) ||
+			  (source.get() && _other.source.get() && source->name() == _other.source->name()));
```
Why not `source.get() == _other.source.get()`?

christianparpart replied:
I'll be doing that in a seperate commit in this PR, as it's also a logic-change (and requires adjusting one test)

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -69,35 +71,41 @@ struct SourceLocation
 		return _out << "NO_LOCATION_SPECIFIED";
-	return _out << *_location.sourceName << "[" << _location.start << "," << _location.end << ")";
+
+	if (_location.source.get())
```
Could be just
```suggestion
	if (_location.source)
```

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -69,35 +71,41 @@ struct SourceLocation
 bool SourceLocation::contains(SourceLocation const& _other) const
 {
-	if (isEmpty() || _other.isEmpty() || ((!sourceName || !_other.sourceName || *sourceName != *_other.sourceName) && (sourceName || _other.sourceName)))
+	if (isEmpty() || _other.isEmpty() || ((!source || !_other.source || source->name() != _other.source->name()) && (source || _other.source)))
```
Again, `source.get() != _outher.source.get()` instead of comparing source names.

chriseth commented on liblangutil/SourceLocation.h:
```diff
@@ -69,35 +71,41 @@ struct SourceLocation
 bool SourceLocation::intersects(SourceLocation const& _other) const
 {
-	if (isEmpty() || _other.isEmpty() || ((!sourceName || !_other.sourceName || *sourceName != *_other.sourceName) && (sourceName || _other.sourceName)))
+	if (isEmpty() || _other.isEmpty() || ((!source || !_other.source || source->name() != _other.source->name()) && (source || _other.source)))
```
Here, too.

Commit:
liblangutil: SourceLocation: Retricts == and != operator
